txp/Pos/StateLock/PrioLock.hs

```diff
@@ -0,0 +1,77 @@
+{-|
+Module:      Pos.StateLock.PrioLock
+Description: Provides a prioritised lock
+
+Provides a lock that can be taken with either high or low precedence.
+Within each precedence, the lock is taken in FIFO order.
+-}
+
+{-# LANGUAGE LambdaCase #-}
+
+module Pos.StateLock.PrioLock where
+
+import           Control.Concurrent.STM      (TMVar, newEmptyTMVar
+                                             , putTMVar, takeTMVar)
+import           Control.Monad.Catch         (MonadMask)
+import           Universum
+
+newtype PrioLock = PrioLock (TVar PrioLockState)
```

Comment:
Naming: I'd make it longer and explicit `PriorityLock`. Clearer and only 4 chars longer. Then update the other names to match.

Reply:
Makes sense!